### PR TITLE
[model compiler] Refactor move-model to be independent of bytecode

### PR DESCRIPTION
**Pull request by wrwg**
This is a large refactor of the `move-model` to become independent of bytecode, and an essential step towards the new compiler.

Originally the move-model was designed to work from bytecode as generated by the Move compiler. As such it was input the prover, docgen, and other tools. This PR makes this dependency optional:

- The model can still operate base off bytecode, supporting all exisiting use cases and tests. This is important for the incremental introduction of the new compiler infra.
- In a new execution path enabled by this PR, the model can also operate without having bytecode as input. In this case, the model builder will completely compile the model from the Move sources.

The focus of this PR is preparing the 2nd operation mode without breaking the first. There are further steps needed for making the 2nd mode fully functional.

This PR also removes a number of unused crates to simplify refactoring:

- stackless bytecode interpreter
- legacy read-write set analysis
- legacy escape analysis

### Test Plan

Existing tests

### CODE REVIEW

File: third_party/move/move-model/tests/testsuite.rs [hunk 1]
```diff
@@ -3,22 +3,16 @@
...
-};
+use move_model::{options::ModelBuilderOptions, run_model_builder_with_options};
 use move_prover_test_utils::baseline_test::verify_or_update_baseline;
 use std::path::Path;
 
 fn test_runner(
     path: &Path,
     options: ModelBuilderOptions,
-    check_from_bytecode: bool,
+    _check_from_bytecode: bool,
```
Comment (davidiw) [inline third_party/move/move-model/tests/testsuite.rs:15]: why was this removed and what are the implications?

File: third_party/move/move-model/src/lib.rs [hunk 2]
```diff
@@ -317,54 +314,6 @@ fn collect_related_modules_recursive<'a>(
     }
 }
 
-/// Build a `GlobalEnv` from a collection of `CompiledModule`'s. The `modules` list must be
-/// topologically sorted by the dependency relation (i.e., a child node in the dependency graph
-/// should appear earlier in the vector than its parents).
-pub fn run_bytecode_model_builder<'a>(
```
Comment (davidiw) [inline third_party/move/move-model/src/lib.rs:323]: what was the motivation in removing this?

**Review by davidiw — COMMENTED** [review]
This is amazing step forward!

Could we consider breaking this up into a few commits? Maybe others have deep familiarity of the code base and can swiftly cut through the noise, but I can't. Right now, I can see at least 5 different distinct pieces that I'd love to be able to review as independently as possible: 
* removal of stackless bytecode interpreter
* removal of legacy read-write set analysis
* removal of legacy escape analysis
* some updates of documentation and other clean ups
* the actual refactor of the move-model

Maybe others will have more time to dig through this.

**Comment by wrwg** [comment]
> This is amazing step forward!
> 
> Could we consider breaking this up into a few commits? Maybe others have deep familiarity of the code base and can swiftly cut through the noise, but I can't. Right now, I can see at least 5 different distinct pieces that I'd love to be able to review as independently as possible:
> 
> * removal of stackless bytecode interpreter
> * removal of legacy read-write set analysis
> * removal of legacy escape analysis
> * some updates of documentation and other clean ups
> * the actual refactor of the move-model
> 
> Maybe others will have more time to dig through this.

The removal of full crates doesn't add mental overhead, so I see no practical reason why to split those parts off.

The other things are unfortunately interconnected and cannot be easily serialized. However, this is really just a refactor. There is no deep review needed IMO. Understanding this PR does not contribute to understanding the new compiler.

**Comment by wrwg** [comment]
> > This is amazing step forward!
> > Could we consider breaking this up into a few commits? Maybe others have deep familiarity of the code base and can swiftly cut through the noise, but I can't. Right now, I can see at least 5 different distinct pieces that I'd love to be able to review as independently as possible:
> > 
> > * removal of stackless bytecode interpreter
> > * removal of legacy read-write set analysis
> > * removal of legacy escape analysis
> > * some updates of documentation and other clean ups
> > * the actual refactor of the move-model
> > 
> > Maybe others will have more time to dig through this.
> 
> The removal of full crates doesn't add mental overhead, so I see no practical reason why to split those parts off.

Forgot to note that those crates are removed because otherwise their code would have to be refactored as well. By removing dead code, I tried to reduce complexity of the refactor.

> 
> The other things are unfortunately interconnected and cannot be easily serialized. However, this is really just a refactor. There is no deep review needed IMO. Understanding this PR does not contribute to understanding the new compiler.

File: third_party/move/move-model/tests/testsuite.rs (hunk 1 above)
Comment (wrwg) [reply third_party/move/move-model/tests/testsuite.rs:15]: This is dead code (only used in tests), and it would have been rather difficult to keep it working.

File: third_party/move/move-model/src/lib.rs (hunk 2 above)
Comment (wrwg) [reply third_party/move/move-model/src/lib.rs:323]: > 

Same as above. Dead code.

**Comment by gerben-stavenga** [comment]
I see you did refactor move-to-yul is that still a package we need to maintain

File: aptos-move/framework/aptos-stdlib/sources/copyable_any.spec.move [hunk 3]
```diff
@@ -4,7 +4,7 @@ spec aptos_std::copyable_any {
     // Function specifications
     // -----------------------
 
-    spec pack<T: drop + store>(x: T): Any {
+    spec pack<T: drop + store + copy>(x: T): Any {
```
Comment (gerben-stavenga) [inline aptos-move/framework/aptos-stdlib/sources/copyable_any.spec.move:7]: Why did this change happen in this PR?

File: third_party/move/move-model/src/ast.rs [hunk 4]
```diff
@@ -1188,30 +1225,27 @@ impl ModuleName {
     /// Determine whether this is a script. The move-compiler infrastructure uses MAX_ADDR
     /// for pseudo modules created from scripts, so use this address to check.
     pub fn is_script(&self) -> bool {
-        static MAX_ADDR: Lazy<BigUint> = Lazy::new(|| {
-            BigUint::from_str_radix("ffffffffffffffffffffffffffffffff", 16).expect("valid hex")
-        });
-        self.0 == *MAX_ADDR
+        self.0 == Address::Numerical(AccountAddress::new([0xFF; AccountAddress::LENGTH]))
```
Comment (gerben-stavenga) [inline third_party/move/move-model/src/ast.rs:1228]: Perhaps make MAX_ADDR  part of AccountAddress

Comment (gerben-stavenga) [reply third_party/move/move-model/src/ast.rs:1228]: Also I thought address was 32 bytes, I count 32 f's in the previous version which is 16 bytes. Was that a bug?

**Review by gerben-stavenga — APPROVED** [review]

File: aptos-move/framework/aptos-stdlib/doc/math128.md [hunk 5]
```diff
@@ -10,12 +10,14 @@ Standard math utilities missing in the Move Language.
 -  [Function `max`](#0x1_math128_max)
 -  [Function `min`](#0x1_math128_min)
 -  [Function `average`](#0x1_math128_average)
+-  [Function `mul_div`](#0x1_math128_mul_div)
```
Comment (brmataptos) [inline aptos-move/framework/aptos-stdlib/doc/math128.md:13]: These doc changes are unrelated to the refactoring, right?

File: aptos-move/framework/aptos-stdlib/sources/copyable_any.spec.move (hunk 3 above)
Comment (wrwg) [reply aptos-move/framework/aptos-stdlib/sources/copyable_any.spec.move:7]: This is because abilities were wrongly implemented before. You could list here whatever abilities you want before instead of those the original fun decl had. The refactor fixed this (kind of unintentional) and so I needed to also correct the source here.

File: third_party/move/move-model/src/ast.rs (hunk 4 above)
Comment (wrwg) [reply third_party/move/move-model/src/ast.rs:1228]: Yes, it was in fact a bug, someone reported this in the Move repo. It only effected docgen for scripts however, which I think no one uses.

File: aptos-move/framework/aptos-stdlib/doc/math128.md (hunk 5 above)
Comment (wrwg) [reply aptos-move/framework/aptos-stdlib/doc/math128.md:13]: No actually not. All inline functions have become visible in the model, because the models function tables are now derived from the parsed AST, and not the bytecode. In the bytecode, inlined functions are gone, so there were not visible in the model before (and hence docgen).

Comment (wrwg) [reply aptos-move/framework/aptos-stdlib/doc/math128.md:13]: So basically -- and unintentionally -- this PRC fixes a known incompleteness of docgen. It similar fixes some link decoration because the parsing of addresses had bugs which are fixed with this PR.

All those changes are _not_ piggybacked on this PR but a consequence of it.

File: aptos-move/framework/aptos-stdlib/sources/copyable_any.spec.move (hunk 3 above)
Comment (gerben-stavenga) [reply aptos-move/framework/aptos-stdlib/sources/copyable_any.spec.move:7]: I suspected as much, could you add it to the description which bugs were fixed?

File: third_party/move/move-model/src/ast.rs (hunk 4 above)
Comment (gerben-stavenga) [reply third_party/move/move-model/src/ast.rs:1228]: Thx, please at to description as fixed as well.

Comment (gerben-stavenga) [reply third_party/move/move-model/src/ast.rs:1228]: Maybe close https://github.com/aptos-labs/aptos-core/issues/7731

File: third_party/move/move-prover/bytecode/src/lib.rs [hunk 6]
```diff
@@ -50,6 +46,9 @@ pub mod verification_analysis;
 pub mod verification_analysis_v2;
 pub mod well_formed_instrumentation;
 
+/// An error message used for cases where a compiled module is expected to be attached
+pub(crate) const COMPILED_MODULE_AVAILABLE: &str = "compiled module missing";
```
Comment (meng-xu-cs) [inline third_party/move/move-prover/bytecode/src/lib.rs:50]: Just a comment that this error message marks a significant portion of the refactoring and can be used to trace where a compiled module is expected in the code.

The name and the static string seems a bit out-of-sync at first sight, but when seeing how it is used, i.e., in `.expect(COMPILED_MODULE_AVAILABLE)`, the naming makes perfect sense.

**Review by meng-xu-cs — APPROVED** [review]
Thanks a lot for this huge effort! I understand this is a large PR but I don't see a way of making it in smaller pieces. The compiled module expectation is a general assumption in the old code and removing this assumption will affect a lot of things.

Fortunately, this is mostly a refactoring to make the Rust type system happy (mainly `Option<T>` to `T` or vice versa) and does not intend to have an impact on the functionality. Giving a green light to the PR based on this.
